Vector/Model/RoomList/RecentsDataSource.m

```diff
@@ -32,10 +47,305 @@ - (instancetype)init
+    
+    if (section == favoritesPos)
+    {
+        count = favoritesCells.count;
+    }
+    else if (section == conversationPos)
+    {
+        count = conversationCells.count;
+    }
+    else if (section == lowPriorityPos)
+    {
+        count = lowPriorityCells.count;
+    }
+
+    return count;
+}
+
+- (UIView *)viewForHeaderInSection:(NSInteger)section withFrame:(CGRect)frame
+{
+    // add multi accounts section management
+    
+    if ((section == favoritesPos) || (section == conversationPos) || (section == lowPriorityPos))
+    {
+        UILabel* label = [[UILabel alloc] initWithFrame:frame];
+        
+        NSString* text = @"";
+        
+        if (section == favoritesPos)
+        {
+            text = NSLocalizedStringFromTable(@"room_recents_favourites", @"Vector", nil);
+        }
+        else if (section == conversationPos)
+        {
+            text = NSLocalizedStringFromTable(@"room_recents_conversations", @"Vector", nil);
+        }
+        else if (section == lowPriorityPos)
+        {
+            text = NSLocalizedStringFromTable(@"room_recents_low_priority", @"Vector", nil);
+        }
+    
+        label.text = [NSString stringWithFormat:@"   %@", text];
+        label.backgroundColor = [UIColor lightGrayColor];
+        
+        return label;
+    }
+    
+    return [super viewForHeaderInSection:section withFrame:frame];
+}
+
+- (id<MXKRecentCellDataStoring>)cellDataAtIndexPath:(NSIndexPath *)indexPath
+{
+    id<MXKRecentCellDataStoring> cellData = nil;
+    
+    if (indexPath.section == favoritesPos)
+    {
+        cellData = [favoritesCells objectAtIndex:indexPath.row];
+    }
+    else if (indexPath.section == conversationPos)
+    {
+        cellData = [conversationCells objectAtIndex:indexPath.row];
+    }
+    else if (indexPath.section == lowPriorityPos)
+    {
+        cellData = [lowPriorityCells objectAtIndex:indexPath.row];
+    }
+    
+    return cellData;
+}
+
+- (CGFloat)cellHeightAtIndexPath:(NSIndexPath *)indexPath
+{
+    id<MXKRecentCellDataStoring> cellData = [self cellDataAtIndexPath:indexPath];
+    
+    if (cellData)
+    {
+        Class<MXKCellRendering> class = [self cellViewClassForCellIdentifier:kMXKRecentCellIdentifier];
+        return [class heightForCellData:cellData withMaximumWidth:0];
+    }
+
+    return 0;
+}
+
+#pragma mark - MXKDataSourceDelegate
+
+// create an array filled with NSNull and with the same size as sourceArray
+- (NSMutableArray*)createEmptyArray:(NSArray*)sourceArray
+{
+    NSMutableArray* array = [[NSMutableArray alloc] init];
+    
+    if (sourceArray && sourceArray.count)
+    {
+        for(int i = 0; i < sourceArray.count; i++)
+        {
+            [array addObject:[NSNull null]];
+        }
+    }
+    
+    return array;
+}
+
+// remove the NSNull from an array
+- (void) removeNullItems:(NSMutableArray*)array
+{
+    NSUInteger pos;
+    
+    while((pos = [array indexOfObject:[NSNull null]]) != NSNotFound)
```

Comment:
Note: [array removeObject:] should be used to remove all [NSNull null] occurrences.